python/src/iceberg/io/pyarrow.py

```diff
@@ -0,0 +1,188 @@
+                raise FileExistsError(f"Cannot create file, already exists: {self.location}")
+            output_file = self._filesystem.open_output_stream(self._path)
+        except PermissionError:
+            raise
+        except OSError as e:
+            if e.errno == 13 or "AWS Error [code 15]" in str(e):
+                raise PermissionError(f"Cannot create file, access denied: {self.location}")
+            raise  # pragma: no cover - If some other kind of OSError, raise the raw error
+        return output_file
+
+    def to_input_file(self) -> "PyArrowFile":
+        """Returns a new PyArrowFile for the location of an existing PyArrowFile instance
+
+        This method is included to abide by the OutputFile abstract base class. Since this implementation uses a single
+        PyArrowFile class (as opposed to separate InputFile and OutputFile implementations), this method effectively returns
+        a copy of the same instance.
+        """
+        return self
+
+
+class PyArrowFileIO(FileIO):
+    def new_input(self, location: str) -> PyArrowFile:
+        """Get a PyArrowFile instance to read bytes from the file at the given location
+
+        Args:
+            location(str): A URI or a path to a local file
+
+        Returns:
+            PyArrowFile: A PyArrowFile instance for the given location
+        """
+        return PyArrowFile(location)
+
+    def new_output(self, location: str) -> PyArrowFile:
+        """Get a PyArrowFile instance to write bytes to the file at the given location
+
+        Args:
+            location(str): A URI or a path to a local file
+
+        Returns:
+            PyArrowFile: A PyArrowFile instance for the given location
+        """
+        return PyArrowFile(location)
+
+    def delete(self, location: Union[str, InputFile, OutputFile]) -> None:
+        """Delete the file at the given location
+
+        Args:
+            location(str, InputFile, OutputFile): The URI to the file--if an InputFile instance or an
+            OutputFile instance is provided, the location attribute for that instance is used as the location
+            to delete
+
+        Raises:
+            FileNotFoundError: When the file at the provided location does not exist
+            PermissionError: If the file at the provided location cannot be accessed due to a permission error such as
+                an AWS error code 15
+        """
+        str_path = location.location if isinstance(location, (InputFile, OutputFile)) else location
+        filesystem, path = FileSystem.from_uri(str_path)  # Infer the proper filesystem
+        try:
+            filesystem.delete_file(path)
+        except OSError as e:
+            if e.errno == 2 or "Path does not exist" in str(e):
+                raise FileNotFoundError(f"Cannot delete file, does not exist: {location}")
+            elif e.errno == 13 or "AWS Error [code 15]" in str(e) or isinstance(e, PermissionError):
```

Comment:
@samredai, here's one that wasn't updated to a separate except clause for `PermissionError`.